File: scripts/evaluate_f5_demo.py
"""Evaluate an F5-TTS demo index with leave-one-out retrieval metrics."""
import argparse
import json
import os
import sqlite3
from datetime import datetime, timezone
from pathlib import Path

import numpy as np


ROOT = Path(__file__).resolve().parents[1]


def category(item):
    return Path(item["path"]).parent.name


def cosine_distance(left, right):
    left, right = np.asarray(left), np.asarray(right)
    return 1 - float(left @ right / max(float(np.linalg.norm(left) * np.linalg.norm(right)), 1e-12))


def prosody_distance(left, right):
    weights = np.asarray([1.4, 1.2, .8, .5, .7, .35, .15])
    scales = np.asarray([.35, .8, .8, .08, .12, .12, 20])
    return float(np.sqrt(np.sum(weights * ((np.asarray(left) - right) / scales) ** 2) / weights.sum()))


def evaluate(name, queries, candidates_for, label_of, vector_of, feature_source, metric, ks=(1, 3, 5)):
    hits = {k: 0 for k in ks}
    reciprocal, failures = [], []
    for query in queries:
        ranked = sorted(
            ((item, metric(vector_of(query), vector_of(item))) for item in candidates_for(query) if item["id"] != query["id"]),
            key=lambda pair: pair[1],
        )
        expected = label_of(query)
        found = next((index for index, (item, _) in enumerate(ranked) if label_of(item) == expected), -1)
        for k in ks:
            hits[k] += 0 <= found < k
        reciprocal.append(1 / (found + 1) if found >= 0 else 0)
        if found != 0:
            failures.append({"query": query["name"], "expected": expected, "top": [
                {"name": item["name"], "label": label_of(item), "distance": round(score, 4)}
                for item, score in ranked[:3]
            ]})
    count = len(queries)
    return {"name": name, "featureSource": feature_source, "queries": count,
            **{f"recallAt{k}": hits[k] / count if count else 0 for k in ks},
            "mrr": sum(reciprocal) / count if count else 0, "failures": failures[:12]}


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--db", default=os.environ.get("DB_PATH", ROOT / ".data" / "f5-baseline.db"))
    parser.add_argument("--audio-root", default=ROOT / "test" / "f5-tts-demo")
    parser.add_argument("--output", default=ROOT / ".data" / "f5-eval-report.json")
    args = parser.parse_args()
    connection = sqlite3.connect(args.db)
    connection.row_factory = sqlite3.Row
    audio_root = str(Path(args.audio_root).resolve())
    items = [{**dict(row), "features": json.loads(row["features_json"])} for row in connection.execute("SELECT * FROM audio_items")
             if row["path"].startswith(audio_root)]
    connection.close()
    has_style = bool(items) and all(item["features"].get("style", {}).get("embedding") for item in items)
    has_emotion = bool(items) and all(item["features"].get("emotion", {}).get("embedding") for item in items)
    prosody = lambda item: item["features"]["prosody"]["vector"]
    style = (lambda item: item["features"]["style"]["embedding"]) if has_style else prosody
    emotion_vector = (lambda item: item["features"]["emotion"]["embedding"]) if has_emotion else prosody
    style_metric = cosine_distance if has_style else prosody_distance
    emotion_metric = cosine_distance if has_emotion else prosody_distance
    emotion = [item for item in items if category(item) == "emotion"]
    robustness = [item for item in items if category(item) == "robustness"]
    speed = [item for item in items if category(item) == "speed_control" and re_search_speed(item["name"])]
    zero_shot = [item for item in items if category(item) == "zero_shot"]
    speed_family = lambda item: re_strip_speed(item["name"])
    report = {
        "generatedAt": datetime.now(timezone.utc).isoformat(), "database": str(args.db),
        "featureVersion": sorted({item["feature_version"] for item in items}), "audioCount": len(items),
        "evaluations": [
            evaluate("emotion-label", emotion, lambda _: emotion, lambda item: item["name"].split("_")[0],
                     emotion_vector, "indextts2-emovec" if has_emotion else "prosody-v1", emotion_metric),
            evaluate("robustness-ref-gen-pair", robustness, lambda _: robustness,
                     lambda item: item["name"].replace("_ref.wav", "").replace("_gen.wav", ""),
                     style, "indextts2-condition" if has_style else "prosody-v1", style_metric),
            evaluate("speed-factor-cross-content", speed, lambda query: [item for item in speed if speed_family(item) != speed_family(query)],
                     lambda item: re_search_speed(item["name"]), style,
                     "indextts2-condition" if has_style else "prosody-v1", style_metric),
            evaluate("zero-shot-reference-group", zero_shot, lambda _: zero_shot,
                     lambda item: item["name"].split("_gen_")[0].removesuffix(".wav"), style,
                     "indextts2-condition" if has_style else "prosody-v1", style_metric),
        ],
    }
    output = Path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(report, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    for result in report["evaluations"]:
        print(f"{result['name']} [{result['featureSource']}]: n={result['queries']} "
              f"R@1={result['recallAt1']:.1%} R@3={result['recallAt3']:.1%} MRR={result['mrr']:.3f}")
    print(f"Report: {output}")


def re_search_speed(name):
    import re
    match = re.search(r"_(0\.7|1\.0|1\.3)x\.wav$", name, re.IGNORECASE)
    return match.group(1) if match else None


def re_strip_speed(name):
    import re
    return re.sub(r"_(0\.7|1\.0|1\.3)x\.wav$", "", name, flags=re.IGNORECASE)


if __name__ == "__main__":
    main()

File: scripts/evaluate_f5_demo.mjs
import path from "node:path";
import fs from "node:fs";

const root = path.resolve("test/f5-tts-demo");
process.env.DB_PATH ||= path.resolve(".data/f5-baseline.db");

const { closeDb, listAudio } = await import("../src/db.js");
const { distance } = await import("../src/search.js");

const items = listAudio().filter(item => item.path.startsWith(root));

function category(item) {
  return path.basename(path.dirname(item.path));
}

function cosineDistance(a, b) {
  let dot = 0, aa = 0, bb = 0;
  for (let i = 0; i < Math.min(a.length, b.length); i++) { dot += a[i] * b[i]; aa += a[i] * a[i]; bb += b[i] * b[i]; }
  return 1 - dot / Math.max(Math.sqrt(aa * bb), 1e-12);
}

function rank(query, candidates, vectorOf, metric) {
  const queryVector = vectorOf(query);
  return candidates
    .filter(item => item.id !== query.id)
    .map(item => ({ item, score: metric(queryVector, vectorOf(item)) }))
    .sort((a, b) => a.score - b.score);
}

function evaluate(name, queries, candidatesFor, labelOf, vectorOf, featureSource, metric, ks = [1, 3, 5]) {
  const hits = Object.fromEntries(ks.map(k => [k, 0]));
  const reciprocalRanks = [];
  const failures = [];
  for (const query of queries) {
    const expected = labelOf(query);
    const ranked = rank(query, candidatesFor(query), vectorOf, metric);
    const foundAt = ranked.findIndex(result => labelOf(result.item) === expected);
    for (const k of ks) if (foundAt >= 0 && foundAt < k) hits[k]++;
    reciprocalRanks.push(foundAt >= 0 ? 1 / (foundAt + 1) : 0);
    if (foundAt !== 0) failures.push({
      query: query.name,
      expected,
      top: ranked.slice(0, 3).map(result => ({ name: result.item.name, label: labelOf(result.item), distance: Number(result.score.toFixed(4)) }))
    });
  }
  return {
    name,
    featureSource,
    queries: queries.length,
    ...Object.fromEntries(ks.map(k => [`recallAt${k}`, queries.length ? hits[k] / queries.length : 0])),
    mrr: queries.length ? reciprocalRanks.reduce((a, b) => a + b, 0) / queries.length : 0,
    failures: failures.slice(0, 12)
  };
}

const prosody = item => item.features.prosody.vector;
const style = item => item.features.style?.embedding;
const emotionVector = item => item.features.emotion?.embedding;
const emotion = items.filter(item => category(item) === "emotion");
const robustness = items.filter(item => category(item) === "robustness");
const speed = items.filter(item => category(item) === "speed_control" && /_(0\.7|1\.0|1\.3)x\.wav$/i.test(item.name));
const zeroShot = items.filter(item => category(item) === "zero_shot");

const speedFamily = item => item.name.replace(/_(0\.7|1\.0|1\.3)x\.wav$/i, "");
const hasStyle = items.length > 0 && items.every(item => style(item)?.length);
const hasEmotion = items.length > 0 && items.every(item => emotionVector(item)?.length);
const styleVector = hasStyle ? style : prosody;
const selectedEmotionVector = hasEmotion ? emotionVector : prosody;
const styleMetric = hasStyle ? cosineDistance : distance;
const emotionMetric = hasEmotion ? cosineDistance : distance;
const report = {
  generatedAt: new Date().toISOString(),
  database: process.env.DB_PATH,
  featureVersion: [...new Set(items.map(item => item.feature_version))],
  audioCount: items.length,
  evaluations: [
    evaluate("emotion-label", emotion, () => emotion, item => item.name.split("_")[0], selectedEmotionVector, hasEmotion ? "indextts2-emovec" : "prosody-v1", emotionMetric),
    evaluate("robustness-ref-gen-pair", robustness, () => robustness, item => item.name.replace(/_(ref|gen)\.wav$/i, ""), styleVector, hasStyle ? "indextts2-condition" : "prosody-v1", styleMetric),
    evaluate("speed-factor-cross-content", speed, query => speed.filter(item => speedFamily(item) !== speedFamily(query)), item => item.name.match(/_(0\.7|1\.0|1\.3)x\.wav$/i)?.[1], styleVector, hasStyle ? "indextts2-condition" : "prosody-v1", styleMetric),
    evaluate("zero-shot-reference-group", zeroShot, () => zeroShot, item => item.name.replace(/_gen_(zh|en|code_switch)\.wav$/i, "").replace(/\.wav$/i, ""), styleVector, hasStyle ? "indextts2-condition" : "prosody-v1", styleMetric)
  ]
};

const output = path.resolve(process.argv[2] || ".data/f5-eval-report.json");
fs.mkdirSync(path.dirname(output), { recursive: true });
fs.writeFileSync(output, `${JSON.stringify(report, null, 2)}\n`);
for (const result of report.evaluations) {
  console.log(`${result.name} [${result.featureSource}]: n=${result.queries} R@1=${(result.recallAt1 * 100).toFixed(1)}% R@3=${(result.recallAt3 * 100).toFixed(1)}% MRR=${result.mrr.toFixed(3)}`);
}
console.log(`Report: ${output}`);
closeDb();
